File: benchmarks/haskell/src/Algorithms/Sieve.hs
module Algorithms.Sieve
  ( sieveCount
  , sieveList
  ) where

import Control.Monad (forM_, when)
import Control.Monad.ST (runST)
import Data.Vector.Unboxed (Vector)
import qualified Data.Vector.Unboxed as V
import qualified Data.Vector.Unboxed.Mutable as MV
import Data.Word (Word8)

-- | Sieve of Eratosthenes using mutable unboxed Vector Word8.
-- Matches Rust's Vec<u8> approach.
-- Returns the count of primes up to n.
sieveCount :: Int -> Int
sieveCount n
  | n < 2     = 0
  | otherwise = V.sum $ V.map fromIntegral (sieveVec n)

-- | Returns all primes up to n.
sieveList :: Int -> [Int]
sieveList n
  | n < 2     = []
  | otherwise =
      let vec = sieveVec n
       in [ i | i <- [2..n], vec V.! i == 1 ]

-- | Internal: compute the sieve as a Word8 vector.
sieveVec :: Int -> Vector Word8
sieveVec n = runST $ do
    arr <- MV.replicate (n + 1) (1 :: Word8)
    MV.write arr 0 0
    MV.write arr 1 0
    let limit = floor (sqrt (fromIntegral n :: Double))
    forM_ [2..limit] $ \p -> do
        isPrime <- MV.read arr p
        when (isPrime == 1) $
            forM_ [p*p, p*p+p .. n] $ \multiple ->
                MV.write arr multiple 0
    V.freeze arr

File: rust/src/algorithms/sieve.rs
/// Sieve of Eratosthenes using mutable Vec<u8>.
/// Returns the count of primes up to n.
pub fn sieve_count(n: usize) -> usize {
    if n < 2 {
        return 0;
    }
    let mut is_prime = vec![1u8; n + 1];
    is_prime[0] = 0;
    is_prime[1] = 0;

    let mut p = 2;
    while p * p <= n {
        if is_prime[p] == 1 {
            let mut multiple = p * p;
            while multiple <= n {
                is_prime[multiple] = 0;
                multiple += p;
            }
        }
        p += 1;
    }

    is_prime.iter().map(|&x| x as usize).sum()
}

/// Returns all primes up to n as a Vec.
pub fn sieve_list(n: usize) -> Vec<usize> {
    if n < 2 {
        return vec![];
    }
    let mut is_prime = vec![1u8; n + 1];
    is_prime[0] = 0;
    is_prime[1] = 0;

    let mut p = 2;
    while p * p <= n {
        if is_prime[p] == 1 {
            let mut multiple = p * p;
            while multiple <= n {
                is_prime[multiple] = 0;
                multiple += p;
            }
        }
        p += 1;
    }

    (2..=n).filter(|&i| is_prime[i] == 1).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_count_primes_up_to_10() {
        assert_eq!(sieve_count(10), 4); // 2, 3, 5, 7
    }

    #[test]
    fn should_count_primes_up_to_100() {
        assert_eq!(sieve_count(100), 25);
    }

    #[test]
    fn should_list_primes_up_to_20() {
        assert_eq!(sieve_list(20), vec![2, 3, 5, 7, 11, 13, 17, 19]);
    }

    #[test]
    fn should_handle_edge_cases() {
        assert_eq!(sieve_count(0), 0);
        assert_eq!(sieve_count(1), 0);
        assert_eq!(sieve_count(2), 1);
    }
}
